src/eval.c

```diff
@@ -480,13 +489,69 @@ sds luaCreateFunction(client *c, robj *body) {
     l->body = body;
     l->flags = script_flags;
     sds sha = sdsnewlen(funcname+2,40);
+    l->node = luaScriptsFIFOAdd(c, sha, evalsha);
     int retval = dictAdd(lctx.lua_scripts,sha,l);
     serverAssertWithInfo(c ? c : lctx.lua_client,NULL,retval == DICT_OK);
     lctx.lua_scripts_mem += sdsZmallocSize(sha) + getStringObjectSdsUsedMemory(body);
     incrRefCount(body);
     return sha;
 }
 
+/* Delete a Lua function with the specified sha.
+ *
+ * This will delete the lua function from the lua interpreter and delete
+ * the lua function from server. */
+void luaDeleteFunction(client *c, sds sha) {
+    /* Delete the script from lua interpreter. */
+    char funcname[43];
+    funcname[0] = 'f';
+    funcname[1] = '_';
+    memcpy(funcname+2, sha, 40);
+    funcname[42] = '\0';
+    lua_pushnil(lctx.lua);
+    lua_setfield(lctx.lua, LUA_REGISTRYINDEX, funcname);
+
+    /* Delete the script from server. */
+    dictEntry *de = dictUnlink(lctx.lua_scripts, sha);
+    serverAssertWithInfo(c ? c : lctx.lua_client, NULL, de);
+    luaScript *l = dictGetVal(de);
+    if (l->node) listDelNode(lctx.lua_scripts_lru_list, l->node);
+    lctx.lua_scripts_mem -= sdsZmallocSize(sha) + getStringObjectSdsUsedMemory(l->body);
+    dictFreeUnlinkedEntry(lctx.lua_scripts, de);
+}
+
+/* Users who abuse EVAL will generate a new lua script on each call, which can
+ * consume large amounts of memory over time. Since EVAL is mostly the one that
+ * abuses the lua cache, and these won't have pipeline issues (scripts won't
+ * disappear when EVALSHA needs it and cause failure), we implement script eviction
+ * only for these (not for one loaded with SCRIPT LOAD). Considering that we don't
+ * have many scripts, then unlike keys, we don't need to worry about the memory
+ * usage of keeping a true sorted LRU linked list.
+ *
+ * evalsha is a hint, indicating whether the lua function is added from the EVAL
+ * context or from the SCRIPT LOAD.
+ *
+ * Returns the corresponding node added, which is used to save it in luaScript
+ * and use it for quick removal and re-insertion into an LRU list each time the
+ * script is used. */
+#define FIFO_LIST_LENGTH 500
+listNode *luaScriptsFIFOAdd(client *c, sds sha, int evalsha) {
```

Comment:
We use both the terms lru and fifo, can we pick one? (I prefer LRU, but fine with whatever)
```suggestion
#define LRU_LIST_LENGTH 500
listNode *luaScriptsLRUAdd(client *c, sds sha, int evalsha) {
```